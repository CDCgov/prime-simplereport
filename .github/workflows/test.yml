name: Test

# Run this on pushes to any branch that change a backend file or the workflow definition
on:
  workflow_dispatch: # because sometimes you just want to force a branch to have tests run
  pull_request:
    branches:
      - "**"
  push:
    branches:
      - main

env:
  NODE_VERSION: 14
  JAVA_VERSION: 11
  PROJECT_ROOT: /home/runner/work/prime-simplereport/prime-simplereport # Hardcoding this here because env.WORKSPACE_ROOT gets overridden by one of the steps downstream. We only need this for Sonar.

defaults:
  run:
    working-directory: backend

jobs:
  test:
    runs-on: ubuntu-latest
    services:
      test-db:
        image: postgres:11-alpine
        env:
          POSTGRES_PASSWORD: this_is_a_super_secure_admin_password
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0 # Disable shallow clones so Sonar can have all the data
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Set up postgresql sidecar
        env:
          PGHOST: localhost
          PGPASSFILE: db-setup/.pgpass_ci
          POSTGRES_USER: postgres
          POSTGRES_DB: postgres
          LIB_DIR: db-setup
        run: |
          chmod 0600 $PGPASSFILE
          db-setup/create-db.sh
      - name: Run Java Tests
        env:
          OKTA_TESTING_DISABLEHTTPS: true
          TWILIO_ACCOUNT_SID: ${{secrets.TWILIO_TEST_ACCOUNT_SID }}
          TWILIO_AUTH_TOKEN: ${{secrets.TWILIO_TEST_AUTH_TOKEN }}
        run: ./gradlew jacocoTestReport -PtestDbPort=5432
      - name: Archive Test Results
        uses: actions/upload-artifact@v2
        if: failure()
        with:
          name: test-report
          path: backend/build/reports/tests/test
          retention-days: 7
      - name: Use Node.js ${{env.NODE_VERSION}}
        uses: actions/setup-node@v2.1.5
        with:
          node-version: ${{env.NODE_VERSION}}
      - name: Cache yarn
        uses: actions/cache@v2
        with:
          path: ~/.cache/yarn
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - name: Node setup
        working-directory: ./frontend
        run: yarn install --prefer-offline
      - name: Test
        working-directory: ./frontend
        run: yarn test:ci
      - name: "Function Apps: npm install"
        working-directory: ./ops/services/app_functions/report_stream_batched_publisher/functions
        run: npm ci
      - name: "Function Apps: Test"
        working-directory: ./ops/services/app_functions/report_stream_batched_publisher/functions
        run: npm run coverage
      - name: Sonar analysis
        if: ${{ github.actor != 'dependabot[bot]' }}
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # Needed to get PR information, if any
          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
        run: ./gradlew sonarqube -Dsonar.projectBaseDir=${{ env.PROJECT_ROOT }} --info
  build-jar:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Run Java Build
        run: ./gradlew assemble
  backend-lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Run Style checks
        run: ./gradlew checkstyleMain checkstyleTest spotlessCheck --continue
  frontend-lint-and-build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Use Node.js ${{env.NODE_VERSION}}
        uses: actions/setup-node@v2.1.5
        with:
          node-version: ${{env.NODE_VERSION}}
      - name: Cache yarn
        uses: actions/cache@v2
        with:
          path: ~/.cache/yarn
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - name: Node setup
        working-directory: ./frontend
        run: yarn
      - name: Lint front-end
        working-directory: ./frontend
        run: yarn lint
      - name: Build release
        working-directory: ./frontend
        run: yarn build
