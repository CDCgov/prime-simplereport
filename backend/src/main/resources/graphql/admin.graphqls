# Queries and mutations that are used only by site admins
# All queries and mutations in this file implicitly require site admin access,
# which is enforced in the API not in the schema validator.
extend type Query {
  organizations(identityVerified: Boolean): [Organization!]!
  pendingOrganizations: [PendingOrganization!]!
}
extend type Mutation {
  uploadPatients(patientList: Upload!): String
  resendToReportStream(testEventIds: [ID!]!): Boolean
  createDeviceType(input: CreateDeviceType!): DeviceType
  updateDeviceType(input: UpdateDeviceType!): DeviceType
  addUser( # not actually used in the admin console (currently)
    name: NameInput
    firstName: String
    middleName: String
    lastName: String
    suffix: String
    email: String!
    organizationExternalId: String!
    role: Role!
  ): User
  createOrganization(
    name: String!
    type: String!
    externalId: String!
    testingFacilityName: String!
    cliaNumber: String
    street: String!
    streetTwo: String
    city: String
    county: String
      @deprecated(reason: "county is derived from smarty streets lookup")
    state: String!
    zipCode: String!
    phone: String
    email: String
    orderingProviderName: NameInput
    orderingProviderFirstName: String
    orderingProviderMiddleName: String
    orderingProviderLastName: String
    orderingProviderSuffix: String
    orderingProviderNPI: String
    orderingProviderStreet: String
    orderingProviderStreetTwo: String
    orderingProviderCity: String
    orderingProviderCounty: String
      @deprecated(reason: "county is derived from smarty streets lookup")
    orderingProviderState: String
    orderingProviderZipCode: String
    orderingProviderPhone: String
    deviceTypes: [String]!
    defaultDevice: String!
    adminName: NameInput
    adminFirstName: String
    adminMiddleName: String
    adminLastName: String
    adminSuffix: String
    adminEmail: String!
  ): Organization
  setOrganizationIdentityVerified(
    externalId: String!
    verified: Boolean!
  ): Boolean
  editPendingOrganization(
    orgExternalId: String!
    name: String
    adminFirstName: String
    adminLastName: String
    adminEmail: String
    adminPhone: String
  ): String
  markPendingOrganizationAsDeleted(
    orgExternalId: String!
    deleted: Boolean!
  ): String
  createOrganizationRegistrationLink(
    organizationExternalId: String!
    link: String!
  ): String
  createFacilityRegistrationLink(
    organizationExternalId: String!
    facilityId: ID!
    link: String!
  ): String
  updateRegistrationLink(link: String!, newLink: String!): String
  setRegistrationLinkIsDeleted(link: String, deleted: Boolean!): String
  adminUpdateOrganization(name: String!, type: String!): String
  markFacilityAsDeleted(facilityId: ID!, deleted: Boolean!): String
}
