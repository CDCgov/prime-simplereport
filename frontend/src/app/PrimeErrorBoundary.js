import React from "react";

// import { SeverityLevel } from "@microsoft/applicationinsights-web";
import ErrorPage from "./commonComponents/ErrorPage";
import { getAppInsights } from "./TelemetryService";

export default class PrimeErrorBoundary extends React.Component {
  state = { hasError: false, error: null };

  componentDidCatch(error, info) {
    this.setState({ hasError: true, error });
    const appInsights = getAppInsights();
    if (appInsights) {
      getAppInsights().trackException({
        exception: new Error("PrimeErrorBoundary: " + error),
        severityLevel: 3,
        properties: { componentStack: info.componentStack },
      });
    }
  }

  render() {
    if (this.state.hasError) {
      return React.createElement(ErrorPage);
    }

    return this.props.children;
  }
}
