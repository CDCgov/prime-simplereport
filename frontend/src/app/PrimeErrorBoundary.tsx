import React, { ErrorInfo } from "react";

import ErrorPage from "./commonComponents/ErrorPage";
import { getAppInsights } from "./TelemetryService";
import { getUrl } from "./utils/url";

interface ErrorBoundaryState {
  hasError: boolean;
  redirectToOkta: boolean;
}

interface PrimeErrorBoundaryProps {
  children?: React.ReactNode;
}

export default class PrimeErrorBoundary extends React.Component<
  PrimeErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: any) {
    super(props);
    this.state = { hasError: false, redirectToOkta: false };
  }

  static getDerivedStateFromError(error: Error) {
    if (error.message === "Not authenticated, redirecting to Okta...") {
      return { redirectToOkta: true };
    }
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error: Error, info: ErrorInfo) {
    if (error.message === "Not authenticated, redirecting to Okta...") {
      return;
    }

    const appInsights = getAppInsights();
    if (appInsights) {
      appInsights.trackException({
        exception: new Error("PrimeErrorBoundary: " + error),
        severityLevel: 3,
        properties: { componentStack: info.componentStack },
      });
    }
  }

  render() {
    if (this.state.hasError) {
      return <ErrorPage />;
    }

    if (this.state.redirectToOkta) {
      const url = `${process.env.REACT_APP_OKTA_URL}/oauth2/default/v1/authorize`;
      const clientId = `?client_id=${process.env.REACT_APP_OKTA_CLIENT_ID}`;
      const redirectUri = `&redirect_uri=${encodeURIComponent(getUrl())}`;
      const responseType = `&response_type=${encodeURIComponent(
        "token id_token"
      )}`;
      const scope =
        "&scope=" +
        encodeURIComponent(
          `openid simple_report ${process.env.REACT_APP_OKTA_SCOPE}`
        );
      const nonce = "&nonce=thisisnotsafe";
      const state = "&state=thisisbogus";
      window.location.href =
        url + clientId + redirectUri + responseType + scope + nonce + state;
      return false;
    }

    return this.props.children;
  }
}

export function stripIdTokenFromOktaRedirectUri(uri: string) {
  const regexJWTAsQueryParam = /(?<=#id_token=).*?(?=&token_type=)/;
  return stripIdTokenFromString(regexJWTAsQueryParam, uri);
}

export function stripIdTokenFromOperationName(operationName: string) {
  const regexOperationName = /(?<=#id_token=).*/;
  return stripIdTokenFromString(regexOperationName, operationName);
}

function stripIdTokenFromString(regex: RegExp, string: string) {
  const idTokenFound = string.match(regex);
  if (idTokenFound === null) return string;
  return string.replace(idTokenFound[0], "{ID-TOKEN-OBSCURED}");
}
