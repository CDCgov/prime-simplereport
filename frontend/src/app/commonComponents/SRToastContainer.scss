.Toastify {
  &__toast-container {
    max-width: 40rem;
    width: auto;
  }

  &__toast {
    padding: 0;
  }

  &__close-button {
    height: 1.5rem;
    margin-top: 0.5rem;
    margin-right: 0.5rem;
    opacity: 1;
    width: 1.5rem;

    &:hover {
      opacity: 0.3;
    }

    svg {
      width: 100%;
      height: 100%;
    }
  }

  .sr-alert {
    &--error {
      background-color: #f4e3db;
    }

    &--success {
      background-color: #ecf3ec;

      .usa-alert--success {
        outline: none;
      }
    }
  }

  .usa-alert {
    max-width: 37.5rem;
  }
}
