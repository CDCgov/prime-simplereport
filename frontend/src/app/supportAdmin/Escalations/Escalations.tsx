import React from "react";

import { useSendSupportEscalationMutation } from "../../../generated/graphql";
import Button from "../../commonComponents/Button/Button";
import { showSuccess } from "../../utils/srToast";
import { useDocumentTitle } from "../../utils/hooks";
import { escalationPageTitle } from "../pageTitles";

export const Escalations = () => {
  useDocumentTitle(escalationPageTitle);
  const [escalateToSlack] = useSendSupportEscalationMutation();

  return (
    <div className="prime-home flex-1">
      <div className="grid-container">
        <div className="grid-row">
          <form className="prime-container card-container">
            <div className="usa-card__header">
              <h1 className="font-heading-lg margin-top-0 margin-bottom-0">
                Escalate to engineering
              </h1>
            </div>
            <div className="usa-card__body margin-top-1">
              <div className="grid-row grid-gap margin-top-2">
                <div className="tablet:grid-col">
                  <p className="usa-prose">
                    Use the below button to notify the SimpleReport engineering
                    team about a tier 2 escalation. Make sure you've filled out
                    information about the escalation in{" "}
                    <a
                      target="_blank"
                      href="https://app.smartsheetgov.com/sheets/wwchRWw4jr55WpX8XwC2h9Mr9rVr3Q3hww64pW61" rel="noreferrer"
                    >
                      the Smartsheet tracker
                    </a>
                  </p>
                  <Button
                    className="width-half margin-top-2 submit-button"
                    onClick={async () => {
                      await escalateToSlack();
                      showSuccess(
                        "The engineering team has been notified that there's a new escalation and will be in touch shortly",
                        `Escalation successfully sent`
                      );
                    }}
                    label={"Submit escalation"}
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
