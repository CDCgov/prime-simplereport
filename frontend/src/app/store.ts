import { createStore } from "redux";

import { COVID_RESULTS } from "../app/constants";

const SET_INITIAL_STATE = "SET_INITIAL_STATE";
const UPDATE_ORGANIZATION = "UPDATE_ORGANIZATION";
const UPDATE_FACILITY = "UPDATE_FACILITY";
const SET_PATIENT = "SET_PATIENT";

// this should be the default value for a brand new org
// TODO: get the fields from a schema or something; hard-coded fields are hard to maintain
const initialState = {
  dataLoaded: false,
  organization: {
    name: "",
  },
  facilities: [],
  facility: {
    id: "",
    name: "",
  },
  user: {
    id: "",
    firstName: "",
    middleName: "",
    lastName: "",
    suffix: "",
    email: "",
    permissions: [],
    roleDescription: "",
    isAdmin: false,
  },
  patient: {
    internalId: "",
    firstName: "",
    lastName: "",
    middleName: "",
    birthDate: "",
    isDeleted: false,
    role: "",
    lastTest: {
      dateAdded: "",
      result: COVID_RESULTS.INCONCLUSIVE,
      dateTested: "",
      deviceTypeModel: "",
    },
  },
};

const reducers = (state = initialState, action: any) => {
  switch (action.type) {
    case SET_INITIAL_STATE:
      return {
        ...state,
        ...action.payload,
      };
    case UPDATE_ORGANIZATION:
      return {
        ...state,
        organization: {
          ...state.organization,
          ...action.payload,
        },
      };
    case UPDATE_FACILITY:
      return {
        ...state,
        facility: {
          ...state.facility,
          ...action.payload,
        },
      };
    case SET_PATIENT:
      return {
        ...state,
        patient: {
          ...action.payload,
        },
      };
    default:
      return state;
  }
};

export const setInitialState = (initialState: any) => {
  return {
    type: SET_INITIAL_STATE,
    payload: initialState,
  };
};

export const updateOrganization = (organization: any) => {
  return {
    type: UPDATE_ORGANIZATION,
    payload: organization,
  };
};

export const updateFacility = (facility: any) => {
  return {
    type: UPDATE_FACILITY,
    payload: facility,
  };
};

export const setPatient = (patient: any) => {
  return {
    type: SET_PATIENT,
    payload: patient,
  };
};

const configureStore = () => {
  return createStore(
    reducers,
    initialState,
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION__()
  );
};

export const store = configureStore();

export type RootState = ReturnType<typeof store.getState>;
