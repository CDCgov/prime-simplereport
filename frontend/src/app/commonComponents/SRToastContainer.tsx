import React from "react";
import { ToastContainer } from "react-toastify";

import "react-toastify/dist/ReactToastify.css";
import "./SRToastContainer.scss";

const SRToastContainer = () => {
  const autoClose = 5000;
  const closeButton = true;
  const hideProgressBar = true;
  const limit = 2;
  const position = "bottom-left";

  return (
    <ToastContainer
      autoClose={autoClose}
      closeButton={closeButton}
      hideProgressBar={hideProgressBar}
      limit={limit}
      position={position}
    />
  );
};

export default SRToastContainer;
