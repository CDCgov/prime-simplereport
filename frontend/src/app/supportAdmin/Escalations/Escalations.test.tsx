import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { MockedProvider } from "@apollo/client/testing";

import PrimeErrorBoundary from "../../PrimeErrorBoundary";
import { store } from "../../store";
import { WhoAmIQueryMock } from "../../ReportingApp.test";
import SRToastContainer from "../../commonComponents/SRToastContainer";
import { SendSupportEscalationDocument } from "../../../generated/graphql";

import { Escalations } from "./Escalations";

const SendSupportEscalationMock = {
  request: {
    query: SendSupportEscalationDocument,
  },
  result: {
    data: {},
  },
};

describe("Escalations", () => {
  it("Renders the page", async () => {
    const { container } = render(
      <PrimeErrorBoundary>
        <Provider store={store}>
          <MockedProvider mocks={[WhoAmIQueryMock]}>
            <MemoryRouter>
              <Escalations />
            </MemoryRouter>
          </MockedProvider>
        </Provider>
      </PrimeErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it("Loads a success toast when submission works correctly", async () => {
    render(
      <div>
        <Provider store={store}>
          <MockedProvider mocks={[WhoAmIQueryMock, SendSupportEscalationMock]}>
            <MemoryRouter>
              <Escalations />
            </MemoryRouter>
          </MockedProvider>
        </Provider>
        <SRToastContainer />
      </div>
    );
    const submitButton = await screen.findByText("Submit escalation");
    submitButton.click();
    await screen.findByText("Escalation successfully sent");
  });
});
