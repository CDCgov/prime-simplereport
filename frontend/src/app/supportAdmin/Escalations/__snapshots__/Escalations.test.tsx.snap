// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should show error UI 1`] = `
<div>
  <p>
    Server connection error...
  </p>
</div>
`;

exports[`Escalations Renders the page 1`] = `
<div>
  <div
    class="prime-home flex-1"
  >
    <div
      class="grid-container"
    >
      <div
        class="grid-row"
      >
        <form
          class="prime-container card-container"
        >
          <div
            class="usa-card__header"
          >
            <h1
              class="font-heading-lg margin-top-0 margin-bottom-0"
            >
              Escalate to engineering
            </h1>
          </div>
          <div
            class="usa-card__body margin-top-1"
          >
            <div
              class="grid-row grid-gap margin-top-2"
            >
              <div
                class="tablet:grid-col"
              >
                <p
                  class="usa-prose"
                >
                  Use the below button to notify the SimpleReport engineering team about a tier 2 escalation. Make sure you've filled out information about the escalation in
                   
                  <a
                    href="https://app.smartsheetgov.com/sheets/wwchRWw4jr55WpX8XwC2h9Mr9rVr3Q3hww64pW61"
                    rel="noreferrer"
                    target="_blank"
                  >
                    the Smartsheet tracker
                  </a>
                </p>
                <button
                  class="usa-button width-half margin-top-2 submit-button"
                  type="button"
                >
                  Submit escalation
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
